### PR TITLE
[MRG] Refactor `guess_spatial_dimensions`

### LINKED ISSUE #3396: Deprecate guess_spatial_dimensions
As discussed in #3394, ``color.guess_spatial_dimensions`` is only used in one function ``gaussian.filter``. In addition, this function has a "magics" taste that we decided to limit in the future. #3394 proposed to move the useful code of ``guess_spatial_dimensions`` to the ``_gaussian.py`` module.

### PR DESCRIPTION
**Pull request by Henley13**
## Description
Move `guess_spatial_dimensions` in _gaussian.py.

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests

## References
Closes #3396

## For reviewers
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by Henley13** [comment]
> The original definition might need to be moved/deprecated? Perhaps we want to deprecate the function altogether?

Indeed, #3396 is entitled "Deprecate guess_spatial_dimensions", but the discussion proposes to move the function... As you want :)

Otherwise, in the gaussian filter, the number of dimension would be directly infer by `spatial_dims = image.ndim` and it is mostly used to feed a potential warning message.

**Comment by jni** [comment]
> Indeed, #3396 is entitled "Deprecate guess_spatial_dimensions", but the discussion proposes to move the function...

Does it? My reading of it is that both things should be done. =) The point is, it was actually part of our public API, so it certainly needs to exist there for the next two releases. So my proposed approach is:

- move the function to a *private* function in `gaussian`
- redefine `color.guess_spatial_dimensions` to raise a deprecation warning and then import and call the private function from `gaussian`
- optionally, deprecate the magic guessing behaviour of `gaussian`. But this last step can wait for a later PR.

File: skimage/filters/_gaussian.py [hunk 1]
```diff
@@ -132,3 +131,34 @@ def gaussian(image, sigma=1, output=None, mode='nearest', cval=0,
     image = convert_to_float(image, preserve_range)
     return ndi.gaussian_filter(image, sigma, mode=mode, cval=cval,
                                truncate=truncate)
+
+
+def guess_spatial_dimensions(image):
```
Comment (stefanv) [inline skimage/filters/_gaussian.py]:
```suggestion
def _guess_spatial_dimensions(image):
```

File: skimage/filters/tests/test_gaussian.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import numpy as np
-from skimage.filters._gaussian import gaussian
+from skimage.filters._gaussian import gaussian, guess_spatial_dimensions
```
Comment (stefanv) [inline skimage/filters/tests/test_gaussian.py]:
```suggestion
from skimage.filters._gaussian import gaussian, _guess_spatial_dimensions
```

File: skimage/filters/tests/test_gaussian.py [hunk 3]
```diff
@@ -68,3 +68,17 @@ def test_4d_ok():
...
     assert np.allclose(res.sum(), 1)
+
+
+def test_guess_spatial_dimensions():
+    im1 = np.zeros((5, 5))
+    im2 = np.zeros((5, 5, 5))
+    im3 = np.zeros((5, 5, 3))
+    im4 = np.zeros((5, 5, 5, 3))
+    im5 = np.zeros((5,))
+    testing.assert_equal(guess_spatial_dimensions(im1), 2)
```
Comment (stefanv) [inline skimage/filters/tests/test_gaussian.py]:
```suggestion
    testing.assert_equal(_guess_spatial_dimensions(im1), 2)
```

File: skimage/filters/tests/test_gaussian.py [hunk 4]
```diff
@@ -68,3 +68,17 @@ def test_4d_ok():
...
+    testing.assert_equal(guess_spatial_dimensions(im2), 3)
```
Comment (stefanv) [inline skimage/filters/tests/test_gaussian.py]:
```suggestion
    testing.assert_equal(_guess_spatial_dimensions(im2), 3)
```

File: skimage/filters/tests/test_gaussian.py [hunk 5]
```diff
@@ -68,3 +68,17 @@ def test_4d_ok():
...
+    testing.assert_equal(guess_spatial_dimensions(im3), None)
```
Comment (stefanv) [inline skimage/filters/tests/test_gaussian.py]:
```suggestion
    testing.assert_equal(_guess_spatial_dimensions(im3), None)
```

File: skimage/filters/tests/test_gaussian.py [hunk 6]
```diff
@@ -68,3 +68,17 @@ def test_4d_ok():
...
+    testing.assert_equal(guess_spatial_dimensions(im4), 3)
```
Comment (stefanv) [inline skimage/filters/tests/test_gaussian.py]:
```suggestion
    testing.assert_equal(_guess_spatial_dimensions(im4), 3)
```

File: skimage/filters/tests/test_gaussian.py [hunk 7]
```diff
@@ -68,3 +68,17 @@ def test_4d_ok():
...
+    with testing.raises(ValueError):
+        guess_spatial_dimensions(im5)
```
Comment (stefanv) [inline skimage/filters/tests/test_gaussian.py]:
```suggestion
        _guess_spatial_dimensions(im5)
```

**Comment by sciunto** [comment]
Continued in #4031  Thanks @Henley13

**Pull request closed without merging**